bpo-43416: Add Include/README.rst

erlend-aasland opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-43416](https://bugs.python.org/issue43416) -->
https://bugs.python.org/issue43416
<!-- /issue-number -->

Commit:
Add Include/README.rst

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+3. ``Include/internal/``
+
+
+Include: Limited API and Stable API
```
Stable **ABI**

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+define to exclude functions from the Limited API.  When a new API is
+introduced, it has to be explicitly excluded using
+``#ifndef Py_LIMITED_API``.  If the author forgets it, the function is
+added to be Stable and Limited API by mistake!
```
It doesn't work like that. Include/cpython/ header files are not included if Py_LIMITED_API macro is defined.

erlend-aasland replied:
C&P from https://bugs.python.org/issue35134#msg329060. I removed the last sentence. Let me know if you want to rephrase it.

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+Please start a public discussion before adding new C API to
+``Include/cpython/``.
+
+The API in ``Include/cpython/`` is prefixed with ``_Py``.
```
This is wrong.

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+returned.
+
+Please start a public discussion before adding new C API to
+``Include/cpython/``.
```
I don't think that this is needed. I suggest to only need it for the Include/.

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+==================================
+
+With extern keyword
+-------------------
```
I suggest to move this function at the end.

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+With extern keyword
+-------------------
+
+``Include/internal/`` with ``extern`` contains the C API which *must not
```
```suggestion
``Include/internal/`` API defined with ``extern`` *must not
```

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+
+``Include/internal/`` with ``extern`` contains the C API which *must not
+and can not* be used outside the CPython code base.  Only stdlib
+extensions built as built-in extensions can use it.
```
```suggestion
and can not* be used outside the CPython code base.  Only built-in stdlib
extensions (built with Py_BUILD_CORE_BUILTIN macro defined) can use it.
```

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+and can not* be used outside the CPython code base.  Only stdlib
+extensions built as built-in extensions can use it.
+
+In case of doubt, new C API must be added to the internal C API using
```
```suggestion
In case of doubt, new internal C API must be defined in the internal C API with
```

vstinner commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+With PyAPI_FUNC or PyAPI_DATA
+-----------------------------
+
+``Include/internal/`` with ``PyAPI_FUNC`` or ``PyAPI_DATA`` contains the
```
```suggestion
``Include/internal/`` API defined with ``PyAPI_FUNC`` or ``PyAPI_DATA`` contains the
```

Commit:
Address review

Commit:
Fix one more stable API => ABI

Commit:
Remove duplicate word

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+Include: Limited API and Stable ABI
+===================================
+
+``Include/`` contains the public Limited API and the Stable ABI.  The
```
Clarify that this excludes the *subdirectories* cpython/ and internal/.

Commit:
Avoid using API when talking about a single function or macro

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+===================================
+
+``Include/`` contains the public Limited API and the Stable ABI.  The
+Stable ABI is defined by :pep:`384`.
```
Can you please summarize in 1-2 sentences what the stable ABI is, and how it relates to the limited API? For someone not already familiar with the distinction it may be a bit much to have to read that PEP.

erlend-aasland replied:
Maybe it is sufficient to add a few words about what an ABI is and what an API is? "Stable" and "Limited" are self-explanatory, right? :)

erlend-aasland replied:
I've updated this section with eb050465ed0a8061b00631025d25073bf15438c2. AFAICT, the Limited API is a part of the Stable ABI, and as such defined by PEP 384.

erlend-aasland replied:
I'm not sure how much detail we should put into this README file. Perhaps a longer and more descriptive version could be put into the Developer's Guide.

erlend-aasland replied:
PTAL. I've stolen some wording from PEP 652.

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+Functions stealing references or returning borrowed references *must not*
+be added to the Limited API API or the Stable ABI.  A strong reference
+*must* be returned.  The API in ``Include/`` should not expose or leak
+implementation details.
```
I'd prefix this with a marker like **NOTE:** and maybe indent the whole paragraph. Or perhaps insert a section subheading before it explaining that here are some rules for making additions or changes to the limited API.

PS: "Limited API API".

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+===============================================
+
+``Include/cpython/`` contains the public API that is excluded from the
+Limited and Stable ABI.  :pep:`384` introduced the ``Py_LIMITED_API``
```
```suggestion
Limited API and the Stable ABI.  :pep:`384` introduced the ``Py_LIMITED_API``
```

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+Limited and Stable ABI.  :pep:`384` introduced the ``Py_LIMITED_API``
+define to exclude functions from the Limited API.  When a new API is
+introduced, it has to be explicitly excluded using
+``#ifndef Py_LIMITED_API``.
```
It's unclear from this *where* you have to put this. In the header file? And why? If it's under Include/cpython/, isn't it automatically excluded?

erlend-aasland replied:
I was under the impression that parts of a header file are explicitly excluded from the limited API:

```
$ grep -r Py_LIMITED_API Include/{cpython,internal}
Include/cpython/pyarena.h:#ifndef Py_LIMITED_API
Include/cpython/pyarena.h:#endif /* Py_LIMITED_API */
Include/cpython/pyctype.h:#ifndef Py_LIMITED_API
Include/cpython/pyctype.h:#endif /* !Py_LIMITED_API */
Include/cpython/pyfpe.h:#ifndef Py_LIMITED_API
Include/cpython/pyfpe.h:#endif /* !defined(Py_LIMITED_API) */
Include/cpython/odictobject.h:#ifndef Py_LIMITED_API
Include/cpython/picklebufobject.h:#ifndef Py_LIMITED_API
Include/cpython/picklebufobject.h:#endif /* !Py_LIMITED_API */
Include/cpython/pytime.h:#ifndef Py_LIMITED_API
Include/cpython/pytime.h:#endif /* Py_LIMITED_API */
Include/cpython/initconfig.h:#ifndef Py_LIMITED_API
Include/cpython/initconfig.h:#endif /* !Py_LIMITED_API */
Include/cpython/parser_interface.h:#ifndef Py_LIMITED_API
Include/cpython/parser_interface.h:#endif /* !Py_LIMITED_API */
Include/cpython/pydebug.h:#ifndef Py_LIMITED_API
Include/cpython/pydebug.h:#endif /* Py_LIMITED_API */
Include/internal/pycore_accu.h:#ifndef Py_LIMITED_API
Include/internal/pycore_accu.h:#endif /* !Py_LIMITED_API */
Include/internal/pycore_import.h:#ifndef Py_LIMITED_API
Include/internal/pycore_import.h:#endif /* !Py_LIMITED_API */
Include/internal/pycore_bytes_methods.h:#ifndef Py_LIMITED_API
Include/internal/pycore_bytes_methods.h:#endif /* !Py_LIMITED_API */
```

@vstinner, would you mind clarifying?

erlend-aasland replied:
Resolved: https://github.com/python/cpython/pull/24884#discussion_r596444339

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+be added to ``Include/cpython/``.  A strong reference *must* be
+returned.
+
+See also `bpo-35134 <https://bugs.python.org/issue35134>`_.
```
That's an issue that contains a lot of text. What's the essence of what we can learn from there? Can't you just say what needs to be known here in this README? (And why is that issue still open?)

erlend-aasland replied:
I included it just as a reference. I guess we can just drop that link. I think most of what is needed from that issue is already included here.

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+
+``Include/internal/`` API defined with ``PyAPI_FUNC`` or ``PyAPI_DATA``
+contains the internal API which is exposed for specific use cases like
+debuggers and profilers.
```
Some English grammar issues here.
```suggestion
Functions in ``Include/internal/`` defined with ``PyAPI_FUNC`` or ``PyAPI_DATA``
are internal functions which are exposed only for specific use cases like
debuggers and profilers.
```

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+
+``Include/internal/`` API defined with ``extern`` *must not and can not*
+be used outside the CPython code base.  Only built-in stdlib extensions
+(built with Py_BUILD_CORE_BUILTIN macro defined) can use it.
```
```suggestion
Functions in ``Include/internal/`` defined with ``extern`` *must not and can not*
be used outside the CPython code base.  Only built-in stdlib extensions
(built with Py_BUILD_CORE_BUILTIN macro defined) can use such functions.
```

gvanrossum commented on Include/README.rst:
```diff
@@ -0,0 +1,63 @@
+(built with Py_BUILD_CORE_BUILTIN macro defined) can use it.
+
+In case of doubt, new internal C API must be defined in the internal C
+API with ``extern``.
```
```suggestion
When in doubt doubt, new internal C functions should be defined
in ``Include/internal/`` using ``extern``.
```

Commit:
Address most of Guido's review

Commit:
Cleanup:

- Improve Include/cpython section
- Clarify Stable ABI/Limited API using wording from PEP 652
- Clean up requirements
- Formatting

pitrou commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+``Include/``, excluding the ``cpython`` and ``internal`` subdirectories,
+contains the public Limited API (Application Programming Interface) and
+the Stable ABI (Application Binary Interface).  :pep:`384` defined the
+Stable ABI and the Limited API; two related, but distinct concepts.
```
PEP 384 doesn't seem to make a difference between "stable ABI" and "limited API". What makes you think the contrary?

erlend-aasland replied:
Quoting from PEP 384:
```
The ABI specification falls into two parts: an API specification, specifying what function (groups) are available for use with the ABI, and a linkage specification specifying what libraries to link with.
```

Quoting from the draft PEP 652:
```
PEP 384 and this document deal with the Limited API and the Stable ABI, two related but distinct concepts. In short:

- The Stable ABI is a promise that certain extensions compiled with CPython 3.x will be binary compatible with all subsequent versions of CPython 3.x.
- The Limited API is a subset of CPython's C API that produces such extensions.
```

As you can see, I've stolen the wording from PEP 652.

If this is wrong, can you point me in the right direction?

pitrou replied:
Hmm... well, it is misleading because it conveys the impression that the stable ABI and the limited API contain _different_ functions. Actually, they are entirely identical.

If you want to stress the distinction between those concepts, you could say something like:
> The limited API is a subset of C API designed to guarantee ABI stability accross Python versions on any given platform (hence it is sometimes also referred to as "the stable ABI").

erlend-aasland replied:
The Stable ABI does not guarantee binary compatibility across platforms. I think that is impossible :) Anyway, the Limited API and the Stable ABI are indeed intertwined. Maybe we should tone down the Stable ABI in this text:

```diff
-Include: Limited API and Stable ABI
-===================================
+Include: Limited API
+====================
 
 ``Include/``, excluding the ``cpython`` and ``internal`` subdirectories,
-contains the public Limited API (Application Programming Interface) and
-the Stable ABI (Application Binary Interface).  :pep:`384` defined the
-Stable ABI and the Limited API; two related, but distinct concepts.
-The Stable ABI promises binary compatibility across subsequent versions of
-CPython 3.x, given that the source code restricts itself to the Limited
-API.  The Limited API is a subset of the C API.
+contains the public Limited API (Application Programming Interface).
+The Limited API is a subset of the C API, designed to guarantee ABI
+stability across Python 3 versions, and is defined in :pep:`384`.
```

pitrou replied:
They are not really intertwined, they are the same.
Ok with your wording otherwise.

erlend-aasland replied:
According to PEP 652, they are not. I any case, I think I prefer the shorter text that only mentions the limited API.

serhiy-storchaka replied:
A macro can be in the limited API, but it is not a part of any ABI. An underscored function or variable can be not in the limited API, but be in the stable ABI if it is used in the expansion of a macro from the limited API. We should keep even no longer used private functions if they are in the stable ABI (e.g. `_PyTrash_deposit_object`).

pitrou commented on Include/README.rst:
```diff
@@ -0,0 +1,68 @@
+``Include/cpython/`` contains the public API that is excluded from the
+Limited API and the Stable ABI.  When a new function or macro is
+introduced, it has to be explicitly excluded from the Limited API using
+``#ifndef Py_LIMITED_API``.
```
Why is that the case? If everything under `include/cpython` is excluded from the limited API, then individual `#ifdef`s are entirely pointless.

erlend-aasland replied:
A `grep -r Py_LIMITED_API Include/{cpython,internal}` reveals that there's a lot of such exclusions. If everything under `Include/cpython` (and `Include/internal`) is excluded, how is that enforced and implemented? Such information would be valuable in this README, IMO.

pitrou replied:
To me it just hints that the first sentence above is wrong :-)

erlend-aasland replied:
Quoting PEP 652:
```
During the compilation of applications, the preprocessor macro Py_LIMITED_API must be defined.
Doing so will hide all definitions that are not part of the ABI.
```

pitrou replied:
This quote relates to the compilation of applications. It doesn't relate to the organization of CPython headers.

erlend-aasland replied:
All right, I've gone through all the header files, and header files under `Include/cpython` are completely excluded by `#ifndef Py_LIMITED_API` wrappers, either explicitly in the actual file, or implicitly by guarding the `#include` in the corresponding `Include` header. I'll adjust the text.

pitrou replied:
But do note that, for some reason, there are also non-limited API declarations in `Include`:
https://github.com/python/cpython/blob/master/Include/unicodeobject.h#L141

So "the public API that is excluded from the Limited API" is not only in `Include/cpython`, it's also in `Include`.
Which makes the directory organization completely confusing and pointless IMHO.

erlend-aasland replied:
Yes, that's true. I'll see if I can address this and at the same time keep things short and concise.

Thanks for reviewing, @pitrou.

Commit:
Don't mention the ABI

Commit:
Remove incorrect sentence regarding Py_LIMITED_API usage

Commit:
How to add to the limited API from a given Py version

Commit:
Formatting
